File: core/python/src/aiddl_core/unit_test/unit_test_runner.py

import aiddl_core.parser.parser as aiddl_parser

from aiddl_core.container import Container

from aiddl_core.representation import Sym
from aiddl_core.representation.sym import TRUE
from aiddl_core.representation import List
from aiddl_core.function.uri import EVAL

ASSERT = Sym("#assert")


def run(c, evaluator, verbose):
    n_tests = 0
    n_successful = 0

    tests = c.get_matching_entries(None, ASSERT, None)
    for test in tests:
        p, t = run_single_test(str(test.get_name()), test.get_value(), evaluator, c.fun_reg, verbose)
        n_successful += p
        n_tests += t
    return (n_successful, n_tests)


def run_single_test(label, test, evaluator, freg, verbose):
    if isinstance(test, List):
        n_passed = 0
        n_total = 0
        for sub_test in test:
            p, t = run_single_test(label, sub_test, evaluator, freg, verbose)
            n_passed += p
            n_total += t
        return n_passed, n_total
    else:
        result = evaluator(test)
        if result == TRUE:
            if verbose:
                print("Test %s: ok %s" % (label, test))
            return 1, 1
        else:
            print("Test %s: FAILURE %s" % (label, test))
            print("Result:", result, "->", type(result))
            exit()
            return 0, 1


def run_aiddl_test_file(fname, context=None):
    return run_aiddl_test_files([fname], context=context)


def run_aiddl_test_files(fnames, context=None):
    if context is None:
        c = Container()
    else:
        c = context

    for fname in fnames:
        aiddl_parser.parse(fname, c, ".")

    evaluator = c.fun_reg.get_function(EVAL)
    evaluator.set_container(c)
    result = run(c, evaluator, True)
    print("Test result: %d/%d" % (result[0], result[1]))
    return result

File: core/scala/src/main/scala/org/aiddl/core/scala/tools/UnitTestRunner.scala
package org.aiddl.core.scala.tools

import org.aiddl.core.scala.parser.Parser
import org.aiddl.core.scala.function.Function
import org.aiddl.core.scala.function.Verbose
import org.aiddl.core.scala.representation.Bool
import org.aiddl.core.scala.container.Container
import org.aiddl.core.scala.eval.Evaluator
import org.aiddl.core.scala.representation.Term
import org.aiddl.core.scala.representation.Sym
import org.aiddl.core.scala.representation.Rational
import org.aiddl.core.scala.function.DefaultFunctionUri as D
import org.aiddl.core.scala.representation.Var
import org.aiddl.core.scala.representation.ListTerm
import org.aiddl.core.scala.representation.Num
import org.aiddl.core.scala.tools.Logger

object UnitTestRunner extends Verbose {
    def run( c: Container, eval: Evaluator, verbose: Boolean ): Term = {
        var numSuccess = 0
        var numTests = 0
        
        val tests = c.getMatchingEntries(Var(), Sym("#assert"), Var())

        for ( e <- tests ) {
        val test = e.value
        if ( test.isInstanceOf[ListTerm] ) {
            var i = 1;
            for ( t <- test.asList ) {
            val pass = singleTest(e.name.toString + " " + i, t, eval, verbose)
            numTests += 1
            i += 1
            if ( pass ) numSuccess += 1
            }
        } else {
            val pass = singleTest(e.name.toString(), test, eval, verbose)
            numTests += 1
            if ( pass ) numSuccess += 1
        }
        }

        Num(numSuccess, numTests)
    }

    def singleTest( label: String, test: Term, eval: Evaluator, verbose: Boolean ): Boolean = {
        val result = eval.apply(test)
        val passed = result match { case Bool(v) => v case _ => false }
        if ( passed ) {
            log(1, label + ": ok >>> " + test)
        } else {
            log(1, "---<= RESULT =>---")
            log(1, result.toString())
            log(1, "---<= EVAL =>---")
            eval.setVerbose(2)
            eval.apply(test)
            eval.setVerbose(0)
            log(1, label + ": FAILURE " + test.toString)
            log(1, " -> " + result.toString())
            throw new IllegalStateException("Test failed")
        }
        passed
    }

    def testFiles( fNames: scala.List[String] ): Boolean = {
        val c = new Container()
        val p = new Parser(c)
        Function.loadDefaultFunctions(c)
        for ( fName <- fNames ) {
            p.parseFile(fName)
        }
        val r = UnitTestRunner.run(c, c.getFunctionOrPanic(D.EVAL).asInstanceOf[Evaluator], true)
        r == Rational(1, 1)
    }

    def testFiles( fNames: scala.List[String], c: Container ): Boolean = {
        val p = new Parser(c)
        for ( fName <- fNames ) {
            p.parseFile(fName)
        }
        val r = UnitTestRunner.run(c, c.getFunctionOrPanic(D.EVAL).asInstanceOf[Evaluator], true)
        r == Rational(1, 1)
    }

}